subprojects/core/src/main/java/org/gradle/execution/plan/NodeSets.java

```diff
@@ -84,7 +84,9 @@ public boolean contains(Object o) {
         @Override
         public boolean remove(Object o) {
             if (set.remove(o)) {
-                removeIndex(indexOf(o));
+                int index = indexOf(o);
```

Comment:
❓ 

Is this to handle the case the set and the indices got out of sync (multiple threads modifying the set)? If so, isn't a `ConcurrentModificationException` more appropriate?

None of the added tests exercise this change, right?

Also, did you manage to reproduce the issue with `removeIf`? Or do you think the test cases exercise a similar scenario and that is not necessary?

Reply:
> Is this to handle the case the set and the indices got out of sync (multiple threads modifying the set)?

Yes, though I don't expect it to ever happen in practice. That's why I chose an `assert` instead of the exception.

> If so, isn't a `ConcurrentModificationException` more appropriate?

Well... We might as well have it.